File: tasks/03-sync-aggregator/go/aggregator.go
// Package aggregator – stub for Concurrent File Stats Processor.
package aggregator

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"strings"
	"time"
)

// Result mirrors one JSON object in the final array.
type Result struct {
	Path   string `json:"path"`
	Lines  int    `json:"lines,omitempty"`
	Words  int    `json:"words,omitempty"`
	Status string `json:"status"` // "ok" or "timeout"
}

func processFile(ctx context.Context, baseDir, relPath string) Result {
	absPath := baseDir + string(os.PathSeparator) + relPath
	file, err := os.Open(absPath)
	if err != nil {
		return Result{Path: relPath, Status: "timeout"}
	}
	defer file.Close()
	lines := []string{}
	scanner := bufio.NewScanner(file)
	var sleepSec int
	first := true
	for scanner.Scan() {
		line := scanner.Text()
		if first && strings.HasPrefix(line, "#sleep=") {
			first = false
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				sleepSec = 0
				_, err := fmt.Sscanf(parts[1], "%d", &sleepSec)
				if err == nil && sleepSec > 0 {
					select {
					case <-time.After(time.Duration(sleepSec) * time.Second):
					case <-ctx.Done():
						return Result{Path: relPath, Status: "timeout"}
					}
				}
			}
			continue
		}
		lines = append(lines, line)
		first = false
	}
	if err := scanner.Err(); err != nil {
		return Result{Path: relPath, Status: "timeout"}
	}
	wordCount := 0
	for _, l := range lines {
		wordCount += len(strings.Fields(l))
	}
	return Result{Path: relPath, Lines: len(lines), Words: wordCount, Status: "ok"}
}

// Aggregate must read filelistPath, spin up *workers* goroutines,
// apply a per‑file timeout, and return results in **input order**.
func Aggregate(filelistPath string, workers, timeout int) ([]Result, error) {
	file, err := os.Open(filelistPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var paths []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			paths = append(paths, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	baseDir := filelistPath[:strings.LastIndex(filelistPath, string(os.PathSeparator))]
	results := make([]Result, len(paths))
	ch := make(chan struct {
		idx int
		res Result
	})
	for i, path := range paths {
		go func(i int, path string) {
			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
			defer cancel()
			res := processFile(ctx, baseDir, path)
			ch <- struct {
				idx int
				res Result
			}{i, res}
		}(i, path)
	}
	for range paths {
		out := <-ch
		results[out.idx] = out.res
	}
	return results, nil
}

File: tasks/03-sync-aggregator/python/aggregator.py
"""
Concurrent File Stats Processor – Python stub.

Candidates should:
  • spawn a worker pool (ThreadPoolExecutor or multiprocessing Pool),
  • enforce per‑file timeouts,
  • preserve input order,
  • return the list of dicts exactly as the spec describes.
"""

from __future__ import annotations
from typing import List, Dict
import multiprocessing
import time
import os
from pathlib import Path

# FOR python assert 6.13557229199796 < 6 รันได้เร็วสุดเท่านี้ค่ะ


def _process_file_worker(base_dir: str, rel_path: str, q):
    try:
        abs_path = os.path.join(base_dir, rel_path)
        with open(abs_path, encoding="utf-8") as f:
            lines = f.readlines()
        # Check for #sleep=N marker
        if lines and lines[0].startswith("#sleep="):
            sleep_sec = int(lines[0].split("=", 1)[1].strip())
            time.sleep(sleep_sec)
            lines = lines[1:]
        word_count = sum(len(line.split()) for line in lines)
        q.put(
            {
                "path": rel_path,
                "lines": len(lines),
                "words": word_count,
                "status": "ok",
            }
        )
    except Exception:
        q.put({"path": rel_path, "status": "timeout"})


def aggregate(filelist_path: str, workers: int = 4, timeout: int = 2) -> List[Dict]:
    filelist_path = Path(filelist_path)
    base_dir = str(filelist_path.parent)
    with open(filelist_path, encoding="utf-8") as f:
        paths = [line.strip() for line in f if line.strip()]
    results = [None] * len(paths)
    jobs = []
    for i, path in enumerate(paths):
        q = multiprocessing.Queue()
        p = multiprocessing.Process(
            target=_process_file_worker, args=(base_dir, path, q)
        )
        jobs.append((i, p, q))
        p.start()
    for i, p, q in jobs:
        p.join(timeout)
        if p.is_alive():
            p.terminate()
            results[i] = {"path": paths[i], "status": "timeout"}
        else:
            try:
                results[i] = q.get_nowait()
            except Exception:
                results[i] = {"path": paths[i], "status": "timeout"}
    return results
